pkg/disk/attachdetach_slot.go

```diff
@@ -0,0 +1,80 @@
+package disk
+
+import (
+	"context"
+	"sync"
+)
+
+type AttachDetachSlots struct {
+	nodes       sync.Map
+	slotFactory func() slot
+}
+
+func (a *AttachDetachSlots) GetSlotFor(node string) slot {
+	value, _ := a.nodes.LoadOrStore(node, a.slotFactory())
+	return value.(slot)
+}
+
+type slot interface {
+	AquireDetach(ctx context.Context) error
+	AquireAttach(ctx context.Context) error
+	Release()
+}
+
+type serialSlot struct {
+	detachSlot chan struct{}
+	normalSlot chan struct{}
+}
+
+func (s *serialSlot) AquireDetach(ctx context.Context) error {
+	select {
+	case s.detachSlot <- struct{}{}:
+		return nil
+	case s.normalSlot <- struct{}{}:
```

Comment:
The test cases should meet the following scenarios on the same node
1. Delete the application pod after restarting the controller pod, triggering an uninstallation.
2. After restarting controller pod, delete/create different pods at the same time, trigger mounting/unmounting at the same time.
3. after restarting the controller pod, create three pods at the same time, two mount and one unmount, and the unmounting is completed first.

Reply:
About "restarting the controller". The invariant here is the slot buffer is filled if and only if an attach or detach is in progress. This holds naturally when the process restarted.

I've run this through the official k/k external storage e2e test, with 25 cases in parallel. It works smoothly.